lemur/auth/views.py

```diff
@@ -183,56 +183,95 @@ def retrieve_user_memberships(user_api_url, user_membership_provider, access_tok
     return user, user_membership
 
 
-def create_user_roles(profile):
+def create_user_roles(profile: dict) -> list[str]:
     """Creates new roles based on profile information.
 
     :param profile:
     :return:
     """
     roles = []
 
-    # update their google 'roles'
-    if "googleGroups" in profile:
-        for group in profile["googleGroups"]:
-            role = role_service.get_by_name(group)
-            if not role:
-                role = role_service.create(
-                    group,
-                    description="This is a google group based role created by Lemur",
-                    third_party=True,
-                )
-            if (group != 'admin') and (not role.third_party):
-                role = role_service.set_third_party(role.id, third_party_status=True)
-            roles.append(role)
-    else:
-        current_app.logger.warning(
-            "'googleGroups' not sent by identity provider, no specific roles will assigned to the user."
-        )
+    # We default to pulling in "googleGroups" as that was historically hard coded
+    idp_groups_keys = current_app.config.get("IDP_GROUPS_KEYS", ["googleGroups"])
 
-    role = role_service.get_by_name(profile["email"])
+    for idp_groups_key in idp_groups_keys:
+        if idp_groups_key not in profile:
+            current_app.logger.warning(
+                f"""'{idp_groups_key}' not sent by identity provider for user {profile["email"]}."""
+            )
+            continue
 
-    if not role:
-        role = role_service.create(
-            profile["email"],
-            description="This is a user specific role",
-            third_party=True,
-        )
-    if not role.third_party:
-        role = role_service.set_third_party(role.id, third_party_status=True)
+        if not isinstance(profile[idp_groups_key], list) and all(isinstance(item, str) for item in profile[idp_groups_key]):
+            # Catch instances where roles are not a list of strings
+            current_app.logger.warning(
+                f"""'{idp_groups_key}' sent by identity provider for user {profile["email"]} is not a list of strings."""
+            )
+            continue
+
+        # Take a fixed set of groups/roles and map it to Lemur roles.
+        # If the IDP_GROUPS_TO_ROLES is empty or not set, nothing happens.
+        idp_group_to_role_map = current_app.config.get("IDP_ROLES_MAPPING", {})
+        matched_roles = [
+            idp_group_to_role_map[role] for role in profile.get(idp_groups_key, []) if role in idp_group_to_role_map
+        ]
+        roles.extend(matched_roles)
+
+        # Automatically create and assign roles from the user profile.
+        if current_app.config.get("IDP_ASSIGN_ROLES_FROM_USER_GROUPS", True):
+            idp_roles_description = current_app.config.get("IDP_ROLES_DESCRIPTION",
+                                                           "Identity provider role (generated by Lemur)")
```

Comment:
nit, perhaps helpful to include name of the IDP - `idp_groups_key` in default description here to make the source obvious. This would also get the description closer to earlier hardcoded one ` description="This is a google group based role created by Lemur",`